### ISSUE TITLE
Change the behavior of wxEvtHandler::ProcessPendingEvents or add try-version of it. Correct the documentation

### ISSUE THREAD

**Issue by GeorgyStolbovoy** [wxWidgets/wxWidgets#23651]
### About the behavior of wxEvtHandler::ProcessPendingEvents
I don't understand why ProcessPendingEvents makes the assertion failed if there are no pending events. At the time of calling this function, it is extremely difficult to be sure that the main thread did not have time to process deferred events coming from different places in the program.

In my case, the secondary thread was making a lot of deferred calls with a GUI change for the main thread. There is also an event handler for suspending, in which the main thread waits for the secondary thread to finish and executes its deferred events. The suspension event may come when the secondary thread has not had time to make all its deferred calls, and the primary thread has to process them. In this case, the function will work as it should. But if the interrupt event comes when the second thread has finished and the main thread has managed to process all its calls, there will be a assertion failed.

Here is my code, which I have greatly simplified:
```c++
#include <wx/wxprec.h>

#ifndef WX_PRECOMP
#include <wx/wx.h>
#endif

#include <thread>

class App : public wxApp
{
public:
	bool OnInit() override;
};

class Frame : public wxFrame
{
public:
	Frame();
	wxTimer timer;
	std::thread worker;
};

wxIMPLEMENT_APP(App);

bool App::OnInit()
{
	Frame* frame = new Frame();
	frame->Show(true);
	return true;
}

Frame::Frame() :
	wxFrame(nullptr, wxID_ANY, "ProcessPendingEvents Problem"),
	timer(this)
{
	timer.Start(1500);
	Bind(wxEVT_TIMER, [this](wxTimerEvent& e)
	{
		if (worker.joinable()) worker.join();
		worker = std::thread{[this]()
		{
			CallAfter([]() {OutputDebugString(L"\nHello");});
			std::this_thread::sleep_for(std::chrono::seconds(1));
			CallAfter([]() {OutputDebugString(L" World!");});
                        // If the interrupt event comes when the secondary thread is here
                        // and the main thread has time to execute the lambda from the line above,
                        // then there will be a assertion failed
		}};
	});
	Bind(wxEVT_LEFT_DOWN, [this](wxMouseEvent& e) // The left click invokes suspension
	{
		timer.Stop();
		if (worker.joinable()) worker.join();
		ProcessPendingEvents(); // Here
	});
	Bind(wxEVT_RIGHT_DOWN, [this](wxMouseEvent& e)
	{
		timer.Start(-1);
	});
}
```
Fortunately, I found an undocumented workaround:
```c++
if (!m_pendingEvents->IsEmpty())
    ProcessPendingEvents();
```
But should users do this?

### About the documentation
I was very confused by the description of ProcessPendingEvents: it says there that it executes events from wxEvtHandler::QueueEvent or wxEvtHandler::AddPendingEvent. In my code, I used CallAfter, and I thought that there was no way to explicitly call the execution of deferred lambdas. After all, ProcessPendingEvents works, judging by the documentation, with wxEvent, not with callable objects. I was going to write here about the problem of the lack of an analogue of ProcessPendingEvents for CallAfter, but at the last moment I used ProcessPendingEvents itself. It turned out that it also works with CallAfter. **It should be in the documentation!**

**Comment by vadz** [comment]
I'm sorry, but before looking at this in the details, could you please explain what exactly are you doing and why are you calling `ProcessPendingEvents()` in the first place? I.e. what are you trying to achieve?

**Comment by GeorgyStolbovoy** [comment]
> I'm sorry, but before looking at this in the details, could you please explain what exactly are you doing and why are you calling `ProcessPendingEvents()` in the first place? I.e. what are you trying to achieve?

For me, this function is used to synchronize threads. When the suspension event arrives (I assigned it to the left click for simplicity), the main thread needs to join the secondary thread, and all the deferred calls that the secondary thread made before completion should also be executed by the main thread. These are the invariants, the logic of the program. Having entered the interrupt handler, the main thread can no longer process new pending calls that the secondary thread will have time to make before completion. I need a synchronization function right in the handler so that the main thread can handle pending calls immediately after the secondary thread joins. That's why I use ProcessPendingEvents.
Pending calls from the secondary thread contain code for processing the GUI. As you know, the processing of the GUI in wxWidgets occurs only in the main thread, and if I do not wrap the work with the GUI in CallAfter, the secondary thread will wait on each such function until the primary thread executes it. And the primary thread, in turn, can receive an interrupt event and wait for the secondary thread - this is a deadlock. Therefore, the secondary thread asynchronously sends the called objects to the main thread.
If, after processing the interrupt event, there are unprocessed pending calls, then this will violate the invariants of the program.
Look at the example. Here, the secondary thread in one of the pending calls delegates the installation of the calculated image:
```c++
worker = std::thread{[this]()
	{
		/*...*/
		wxBitmap bitmap{calculated_image};
		CallAfter([this, &bitmap]()
		{
			/* ...some GUI functions... */
			m_static_bitmap->SetBitmap(bitmap);
		});
	}};
```
And now imagine that before this CallAfter, the main thread entered in such interrupt handler:
```c++
void Frame::OnSuspend(wxEvent& e)
{
        /*...*/
	if (worker.joinable()) worker.join();
	m_static_bitmap->SetBitmap(bitmap_placeholder);
        /*...*/
}
```
It is expected that after the secondary thread is joined, a placeholder image will be set, but while the main thread is waiting for the connection, the secondary thread makes a pending call, which has its own picture setting. And after the interrupt handler, the main thread will execute this pending task, replacing the placeholder image. That's so that this does not happen, immediately after joining, synchronization is needed - the execution of all deferred tasks through ProcessPendingEvents.

Perhaps my case is confusing and not easy to understand. But, in my opinion, it is obvious that synchronization by the ProcessPendingEvents function may be required in some cases.
But what I don't understand is why this function does assertion failed. It is clear that it is extremely problematic to catch cases when deferred events will be available and when they are not. In my case, an interruption can come at any time, and at the time of the synchronization place after the secondary thread is joined, there may be a lot of pending events, or there may not be at all. This circumstance does not affect the logic of my program in any way, but for some reason was made a assertion failed out of it in wxWidgets.

**Comment by vadz** [comment]
Sorry, I think we have a fundamental disagreement here. I don't believe `ProcessPendingEvents()` should be used for thread synchronization and, at any rate, I can guarantee that it was never meant to be used for this.

To avoid deadlocks in your example you should only call `thread::join()` when the thread has almost (or already) exited, i.e. typically the worker thread would send an event just before exiting and the main thread would join it then, when it's sure that this call doesn't block. Calling `join()` like you do is definitely a bad idea because it could block for a long time if the worker thread is busy with something.

There are several possibilities to make your example work: you could set some (atomic) variable to indicate to the worker thread that it has been cancelled and so it shouldn't send its result to the main thread (and, if possible, exit a.s.a.p.). Or you could set a flag in the main thread and ignore the result received from the worker thread if it's set. But none of this involves using `ProcessPendingEvents()` and I don't think you could make this work reliably using this function, there would always be race conditions if you tried.

So IMO there is really nothing to do here. If you really, really want to, please submit a patch adding `HasPendingEvents()`, which is missing now and would seem to be logical to have, but I think changing your code would be much more promising.

**Comment by GeorgyStolbovoy** [comment]
@vadz , using atomics is a good idea. I made it so that before each CallAfter, the secondary thread checks atomic_flag and returns if the flag is true. And the main thread after joining calls DeletePendingEvents - I decided that late events are not needed in my program.
Before that, I called ProcessPendingEvents and there was no race, everything worked fine. It could not have been in my case (despite the fact that this function is most likely not thread-safe), because only one thread is running during its call, the second is joined:
```c++
if (worker.joinable()) worker.join();
// there is only one thread left
        ProcessPendingEvents();
```
Thus, the function can be used to communicate between threads. Moreover, pending events are the recommended means for thread interaction in wxWidgets - that's what the documentation says:

> GUI calls, such as those to a wxWindow or wxBitmap are explicitly not safe at all in secondary threads and could end your application prematurely...
...However, the recommended way is to simply process the GUI calls in the main thread through an event that is posted by wxQueueEvent()

ProcessPendingEvents is part of this mechanism.
But back to the original problem: why does this function make assertion failed? Do you have any suggestions? I think you can face this problem without multithreading. It will be necessary to complete all the remaining pending tasks at one moment - and they can already be completed by the thread while it was idle! And what, is this an error that leaves only the assertion failed?
I think HasPendingEvents is unnecessary here - just remove assertion failed from ProcessPendingEvents, instead - just a return.
And I also remind that it should be written in the documentation that this function works with CallAfter.

**Comment by vadz** [comment]
> But back to the original problem: why does this function make assertion failed?

Because it's only supposed to be called when there are some pending events to dispatch. It's even documented as its precondition.

Again, I don't really mind changing this, but in an ideal world we wouldn't expose this function at all because I don't think it's ever a good idea to call it explicitly, it's only used by wx itself. Of course, it's too late to change this now and as it's part of the public API we have to live with it, but my advice is to not use it at all.

**Comment by GeorgyStolbovoy** [comment]
> > But back to the original problem: why does this function make assertion failed?
>
> Because it's only supposed to be called when there are some pending events to dispatch.

But in general, it is impossible to know for sure that the expected pending events will remain at the time of calling this function. Before that, all of them can be executed in the main loop of wxWidgets. This function is a lottery, it is dangerous to use it, and it is unacceptable for a public API. And since ProcessPendingEvents are already in it, the source of danger should be removed

**Comment by vadz** [comment]
As I said, I'm not against adding some `wxEvtHandler::HasPendingEvents()` but I'm not motivated to do it myself as I don't think it's a good idea to call `ProcessPendingEvents()` directly at all (BTW, note that I'm speaking about `wxEvtHandler` version of this function, the one in `wxApp` is fine).

And as I already wrote, ideally it shouldn't have been part of the public API at all, but it's a bit (by 20+ years) too late for this. And I guess there are cases when it's safe to call it, notably if you add pending events from the main thread itself, so it's not like it can't be _ever_ user safely.
